packages/core/src/document.ts

```diff
@@ -113,97 +113,113 @@ class Document extends EventEmitter {
       pinStore: PinStore,
       context: Context,
       opts: DocOpts = {},
-      validate = true
-  ): Promise<Document> {
+      validate = true): Promise<Document> {
     // Fill 'opts' with default values for any missing fields
     opts = {...DEFAULT_LOAD_DOCOPTS, ...opts}
 
-    const doctype = new handler.doctype(null, context) as T
-    const doc = new Document(id, dispatcher, pinStore, validate, context, handler, doctype)
+    const doc = await Document._loadGenesis(id, handler, dispatcher, pinStore, context, validate)
 
-    const record = await dispatcher.retrieveRecord(doc._genesisCid)
+    if (!id.version) {
+      // No version requested, so we should load the most current version we can find.
+      return await Document._syncDocumentToCurrent(doc, pinStore, opts)
+    } else {
+      // Requested document at a specific version
+      return await Document._syncDocumentToVersion(doc, id.version, dispatcher)
+    }
+  }
 
+  /**
+   * Takes a document containing only the genesis record and kicks off the process to load and apply
+   * the most recent Tip to it.
+   * @param doc - Document containing only the genesis record
+   * @param pinStore
+   * @param opts
+   */
+  static async _syncDocumentToCurrent(doc: Document, pinStore: PinStore, opts: DocOpts): Promise<Document> {
+    // TODO: Assert that doc contains only the genesis record
+    const id = doc.id
+
+    // Update document state to cached state if any
     const isStored = await pinStore.stateStore.exists(id)
     if (isStored) {
       doc._doctype.state = await pinStore.stateStore.load(id)
     }
```

Comment:
I think this (`isStored`) also needs to happen in the `_syncDocumentToVersion` method.

Reply:
Well, as written I don't think that works as there's no guarantee that the cached state is on the same branch of history as the version being requested.  I think we can do this as part of https://github.com/ceramicnetwork/js-ceramic/issues/614 though, and I just added a comment there with an approach that I think will work that we can do now.  But I'd like to do that as a follow-up since I have a stack of commits built on top of this one